File: benchmarks/Java/src/havlak/UnionFindNode.java
package havlak;

import som.Vector;


/**
 * class UnionFindNode
 *
 * The algorithm uses the Union/Find algorithm to collapse
 * complete loops into a single node. These nodes and the
 * corresponding functionality are implemented with this class
 */
final class UnionFindNode {

  private UnionFindNode parent;
  private BasicBlock    bb;
  private SimpleLoop    loop;
  private int           dfsNumber;

  UnionFindNode() { }

  // Initialize this node.
  //
  public void initNode(final BasicBlock bb, final int dfsNumber) {
    this.parent     = this;
    this.bb         = bb;
    this.dfsNumber  = dfsNumber;
    this.loop       = null;
  }

  // Union/Find Algorithm - The find routine.
  //
  // Implemented with Path Compression (inner loops are only
  // visited and collapsed once, however, deep nests would still
  // result in significant traversals).
  //
  public UnionFindNode findSet() {
    Vector<UnionFindNode> nodeList = new Vector<>();

    UnionFindNode node = this;
    while (node != node.parent) {
      if (node.parent != node.parent.parent) {
        nodeList.append(node);
      }
      node = node.parent;
    }

    // Path Compression, all nodes' parents point to the 1st level parent.
    nodeList.forEach(iter -> iter.union(parent));
    return node;
  }

  // Union/Find Algorithm - The union routine.
  //
  // Trivial. Assigning parent pointer is enough,
  // we rely on path compression.
  //
  public void union(final UnionFindNode basicBlock) {
    parent = basicBlock;
  }

  // Getters/Setters
  //
  public BasicBlock getBb() {
    return bb;
  }

  public SimpleLoop getLoop() {
    return loop;
  }

  public int getDfsNumber() {
    return dfsNumber;
  }

  public void setLoop(final SimpleLoop loop) {
    this.loop = loop;
  }
}

File: benchmarks/C++HJ/havlak/UnionFindNode.cpp
#include "UnionFindNode.h"

namespace havlak {

    UnionFindNode::UnionFindNode() { } 

    // Initialize this node.
    void UnionFindNode::initNode(shared_ptr<BasicBlock> bb, int dfsNumber) {
        _parent = shared_from_this();
        _bb = bb;
        _dfsNumber = dfsNumber;
        _loop = nullptr;
    }

    // Union/Find Algorithm - The find routine.
    shared_ptr<UnionFindNode> UnionFindNode::findSet() {
        Vector<shared_ptr<UnionFindNode>> nodeList = Vector<shared_ptr<UnionFindNode>>();

        shared_ptr<UnionFindNode> node = shared_from_this();
        while (node != node->_parent) {
            if (node->_parent != node->_parent->_parent) {
                nodeList.append(node);
            }
            node = node->_parent;
        }

        // Path Compression, all nodes' parents point to the 1st level parent.
        nodeList.forEach([&](shared_ptr<UnionFindNode> iter) -> void {
            iter->unionSet(_parent);
        });
        return node;
    }

    // Union/Find Algorithm - The union routine.
    // Trivial. Assigning parent pointer is enough, we rely on path compression.
    void UnionFindNode::unionSet(shared_ptr<UnionFindNode> basicBlock) {
        _parent = basicBlock;
    }

    // Getters/Setters
    shared_ptr<BasicBlock> UnionFindNode::getBb() {
        return _bb;
    }

    shared_ptr<SimpleLoop> UnionFindNode::getLoop() {
        return _loop;
    }

    int UnionFindNode::getDfsNumber() {
        return _dfsNumber;
    }

    void UnionFindNode::setLoop(shared_ptr<SimpleLoop> loop) {
        _loop = loop;
    }
}
